### PR TITLE
eshtri-cli: serve: support for transparenty serving index.html [DEVINFRA-892]

### PR DESCRIPTION
**Pull request by silverjam**
Allows `index.html` to be served when visiting the root of a "directory" in S3.  Falls back to a directory listing if the `index.html` does not exist.

Jira: [DEVINFRA-892]

[DEVINFRA-892]: https://swift-nav.atlassian.net/browse/DEVINFRA-892?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: crates/esthri-cli/src/http_server.rs [hunk 1]
```diff
@@ -357,43 +362,14 @@ async fn create_error_monitor_stream<T: Stream<Item = io::Result<BytesMut>> + Un
     }
 }
 
-// Work around for https://github.com/Majored/rs-async-zip/issues/11 By default,
```
Comment (silverjam) [inline crates/esthri-cli/src/http_server.rs:360]: I believe this was fixed upstream, there is a new release and @samvrlewis's bug was marked fixed: https://github.com/Majored/rs-async-zip/issues/11

File: crates/esthri-cli/src/http_server.rs [hunk 2]
```diff
@@ -526,10 +502,40 @@ fn format_title(bucket: &str, path: &str) -> Markup {
...
+    s3: &T,
+    bucket: &str,
+    path: &str,
+) -> Result<bool, warp::Rejection> {
+    if is_object(s3, bucket, path).await? {
+        return Ok(false);
+    }
+    let mut directory_list_stream = list_directory_stream(s3, bucket, path);
+    match directory_list_stream.try_next().await {
+        Ok(None) => Ok(false),
```
Comment (silverjam) [inline crates/esthri-cli/src/http_server.rs:527]: Since S3 doesn't really have directories, path prefixes aren't considered a directory until there's at least one item returned

File: crates/esthri-cli/Cargo.toml [hunk 3]
```diff
@@ -56,4 +56,4 @@ backtrace = "0.3"
 fs_extra = "1.2"
 md5 = "0.7"
 tempdir = "0.3"
-zip = "0.6.2"
+zip = "0"
```
Comment (silverjam) [inline crates/esthri-cli/Cargo.toml:59]: Weird version conflict forced me to loosen this constraint:

```
error: failed to select a version for `zstd-sys`.
    ... required by package `zstd-safe v4.1.4+zstd.1.5.2`
    ... which satisfies dependency `zstd-safe = "=4.1.4"` of package `zstd v0.10.0+zstd.1.5.2`
    ... which satisfies dependency `zstd = "^0.10.0"` of package `zip v0.6.0`
    ... which satisfies dependency `zip = "^0.6"` of package `esthri-cli v8.4.0 (/home/ubuntu/dev/esthri/crates/esthri-cli)`
versions that meet the requirements `=1.6.3` are: 1.6.3+zstd.1.5.2

the package `zstd-sys` links to the native library `zstd`, but it conflicts with a previous package which links to `zstd` as well:
package `zstd-sys v2.0.1+zstd.1.5.2`
    ... which satisfies dependency `zstd-sys = "=2.0.1"` of package `zstd-safe v5.0.1+zstd.1.5.2`
    ... which satisfies dependency `zstd-safe = "^5.0.1"` of package `async-compression v0.3.14`
    ... which satisfies dependency `async-compression = "^0.3.14"` of package `async_zip v0.0.8`
    ... which satisfies dependency `async_zip = "^0.0.8"` of package `esthri-cli v8.4.0 (/home/ubuntu/dev/esthri/crates/esthri-cli)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='zstd-sys' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `zstd-sys` which could resolve this conflict
[Finished running. Exit status: 101]
```

File: crates/esthri-cli/Cargo.toml [hunk 4]
```diff
@@ -56,4 +56,4 @@ backtrace = "0.3"
 fs_extra = "1.2"
 md5 = "0.7"
 tempdir = "0.3"
-zip = "0.6.2"
+zip = "0.5"
```
Comment (silverjam) [inline crates/esthri-cli/Cargo.toml:59]: Version conflict forced me to downgrade this library:

```
error: failed to select a version for `zstd-sys`.
    ... required by package `zstd-safe v4.1.4+zstd.1.5.2`
    ... which satisfies dependency `zstd-safe = "=4.1.4"` of package `zstd v0.10.0+zstd.1.5.2`
    ... which satisfies dependency `zstd = "^0.10.0"` of package `zip v0.6.0`
    ... which satisfies dependency `zip = "^0.6"` of package `esthri-cli v8.4.0 (/home/ubuntu/dev/esthri/crates/esthri-cli)`
versions that meet the requirements `=1.6.3` are: 1.6.3+zstd.1.5.2

the package `zstd-sys` links to the native library `zstd`, but it conflicts with a previous package which links to `zstd` as well:
package `zstd-sys v2.0.1+zstd.1.5.2`
    ... which satisfies dependency `zstd-sys = "=2.0.1"` of package `zstd-safe v5.0.1+zstd.1.5.2`
    ... which satisfies dependency `zstd-safe = "^5.0.1"` of package `async-compression v0.3.14`
    ... which satisfies dependency `async-compression = "^0.3.14"` of package `async_zip v0.0.8`
    ... which satisfies dependency `async_zip = "^0.0.8"` of package `esthri-cli v8.4.0 (/home/ubuntu/dev/esthri/crates/esthri-cli)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='zstd-sys' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `zstd-sys` which could resolve this conflict
[Finished running. Exit status: 101]
```

File: crates/esthri-cli/src/http_server.rs [hunk 5]
```diff
@@ -114,12 +113,8 @@ impl ErrorTracker {
         }
     }
     fn record_error(error_tracker: ErrorTrackerArc, err: anyhow::Error) {
```
Comment (achubbic-snav) [inline crates/esthri-cli/src/http_server.rs:115]: While touching this function, please add a comment describing the role of the function and the meaning of the parameters.

I have yet to find a standard for Rust documentation. If you are aware of one, or one that we as a team use, would you please link it (here or slack)

File: crates/esthri-cli/src/http_server.rs [hunk 6]
```diff
@@ -224,10 +226,12 @@ fn with_s3_client(
 pub fn esthri_filter(
```
Comment (achubbic-snav) [inline crates/esthri-cli/src/http_server.rs:226]: Please add a comment describing this function and this filters use cases

File: crates/esthri-cli/src/http_server.rs [hunk 7]
```diff
@@ -714,10 +714,50 @@ fn sanitize_filename(filename: String) -> String {
...
+            .status(StatusCode::FOUND)
+            .body(Body::empty())
+            .map_err(|err| EsthriRejection::warp_rejection(format!("{}", err)));
+        Ok((is_dir_listing, Some(response)))
+    } else {
+        Ok((is_dir_listing, None))
+    }
+}
+
+async fn maybe_serve_index_html<'a, T: S3 + Send>(
```
Comment (achubbic-snav) [inline crates/esthri-cli/src/http_server.rs:737]: Please add documentation. Role of function is not clear from name.

File: crates/esthri-cli/src/http_server.rs [hunk 8]
```diff
@@ -714,10 +714,50 @@ fn sanitize_filename(filename: String) -> String {
     )
 }
 
+async fn maybe_redirect<'a, T: S3 + Send>(
```
Comment (achubbic-snav) [inline crates/esthri-cli/src/http_server.rs:717]: Please consider renaming function. e.g. "redirect_on_..."

File: crates/esthri-cli/tests/integration/http_server.rs [hunk 9]
```diff
@@ -100,6 +100,21 @@ fn upload_test_data() -> anyhow::Result<()> {
     Ok(())
 }
 
+async fn upload_index_url_test_data() -> anyhow::Result<()> {
```
Comment (achubbic-snav) [inline crates/esthri-cli/tests/integration/http_server.rs:103]: Please add comment describing what the test aims to validate

File: crates/esthri-cli/tests/integration/http_server.rs [hunk 10]
```diff
@@ -292,3 +307,78 @@ fn test_fetch_archive_with_compressed_files() {
         key_hash_pairs,
     );
 }
+
+#[tokio::test]
```
Comment (achubbic-snav) [inline crates/esthri-cli/tests/integration/http_server.rs:311]: Please add comment describing what the test aims to validate, including whether this addresses nominal behavior or edge cases if applicable.

**Review by achubbic-snav — CHANGES_REQUESTED** [review]
Some minor comments related to commenting and documentation.

File: crates/esthri-cli/src/http_server.rs (hunk 6 above)
Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:226]: done

File: crates/esthri-cli/src/http_server.rs (hunk 8 above)
Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:731]: renamed

File: crates/esthri-cli/src/http_server.rs (hunk 7 above)
Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:751]: done

File: crates/esthri-cli/tests/integration/http_server.rs (hunk 9 above)
Comment (silverjam) [reply crates/esthri-cli/tests/integration/http_server.rs:103]: this isn't a test, just a helper function for populating test data

File: crates/esthri-cli/src/http_server.rs (hunk 5 above)
Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:115]: This is the standard for Rust documentation: https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html

Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:122]: Updated
